pgjdbc/src/test/java/org/postgresql/test/jdbc2/TimezoneCachingTest.java

```diff
@@ -0,0 +1,303 @@
+      assertTrue("Cache should NOT be set", getTimeZoneCache(rs) == null);
+      rs.getInt(1);
+      assertTrue("Cache should NOT be set", getTimeZoneCache(rs) == null);
+      rs.getObject(2);
+      assertTrue("Cache should be set", getTimeZoneCache(rs) != null);
+      rs.close();
+      rs = stmt.executeQuery("SELECT col1, col2 FROM testtz");
+      rs.next();
+      assertTrue("Cache should NOT be set", getTimeZoneCache(rs) == null);
+      rs.getInt(1);
+      assertTrue("Cache should NOT be set", getTimeZoneCache(rs) == null);
+      rs.getDate(2);
+      assertTrue("Cache should be set", getTimeZoneCache(rs) != null);
+      rs.close();
+    } finally {
+      if (null != rs) {
+        rs.close();
+      }
+      if (null != pstmt) {
+        pstmt.close();
+      }
+      if (null != stmt) {
+        stmt.close();
+      }
+      con.rollback();
+    }
+  }
+
+  /**
+   * Test to check the internal cached timezone of a result set is used as expected.
+   */
+  public void testResultSetCachedTimezoneUsage() throws SQLException {
+    Timestamp ts = new Timestamp(2016 - 1900, 1 - 1, 31, 0, 0, 0, 0);
+    Statement stmt = null;
+    PreparedStatement pstmt = null;
+    ResultSet rs = null;
+    TimeZone tz1 = TimeZone.getTimeZone("GMT+8:00");
+    TimeZone tz2 = TimeZone.getTimeZone("GMT-2:00");
+    try {
+      TimeZone.setDefault(tz1);
+      pstmt = con.prepareStatement("INSERT INTO testtz VALUES (?,?)");
+      pstmt.setInt(1, 1);
+      pstmt.setTimestamp(2, ts);
+      pstmt.addBatch();
+      pstmt.executeBatch();
+      stmt = con.createStatement();
+      rs = stmt.executeQuery("SELECT col1, col2 FROM testtz");
+      rs.next();
+      rs.getInt(1);
+      assertTrue("Timestamps sould be in same time zone", rs.getTimestamp(2).equals(ts));
+      rs.close();
+      rs = stmt.executeQuery("SELECT col1, col2 FROM testtz");
+      rs.next();
+      rs.getInt(1);
+      TimeZone.setDefault(tz2);
+      assertTrue("Timestamps sould NOT be in same time zone", !rs.getTimestamp(2).equals(ts));
+      TimeZone.setDefault(tz1);
+      assertTrue("Timestamps sould NOT be in same time zone", !rs.getTimestamp(2).equals(ts));
+      rs.close();
+    } finally {
+      TimeZone.setDefault(null);
+      if (null != rs) {
+        rs.close();
+      }
+      if (null != pstmt) {
+        pstmt.close();
+      }
+      if (null != stmt) {
+        stmt.close();
+      }
+      con.rollback();
+    }
+  }
+
+  private boolean checkTimestamp(Statement stmt, Timestamp ts, TimeZone tz) throws SQLException {
+    TimeZone prevTz = TimeZone.getDefault();
+    TimeZone.setDefault(tz);
+    ResultSet rs = stmt.executeQuery("SELECT col2 FROM testtz");
+    rs.next();
+    Timestamp dbTs = rs.getTimestamp(1);
+    rs.close();
+    TimeZone.setDefault(prevTz);
+    return dbTs.equals(ts);
+  }
+
+  private TimeZone getTimeZoneCache(Object stmt) {
+    try {
+      Field defaultTimeZoneField = stmt.getClass().getDeclaredField("defaultTimeZone");
```

Comment:
I wonder if the test can be written in terms of `TimeZone.setDefault`.
In other words, if the test could examine how pgjdbc handles changes `TimeZone.setDefault` rather than exploring how it stores the cache internally.